### PR TITLE
Design hibernate regions

### PR DESCRIPTION
**Pull request by BusyJay**
This RFC proposes to hibernate a peer when it reaches a stable state to reduce CPU waste.

The proposal here may need to be formalized and verified using TLA+. @nolouch would you like to help me with it?

### CODE REVIEW

File: text/2019-03-04-hibernate-raft.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+## Summary
+
+This RFC proposes to hibernate a peer when it reaches a stable state.
+
+## Motivation
+
+There are several ticks when driving a peer. The cost of tick is usually
+small. However, when there are a lot of regions in one TiKV instances, the
+cost can't be eligible. We have observed that an idle cluster with many
```
Comment (gregwebs) [inline text/2019-03-04-hibernate-raft.md]: eligible -> negligible

File: text/2019-03-04-hibernate-raft.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+regions can waste a lot of CPUs to send useless heartbeats and tick to routine
+checks. Even in an online cluster with tens of thousands of regions, enlarge
+the ticks' interval can usually lead to better performance. Threaded raftstore
+can ease the problem but can't solve it. In general, not all regions keep
+working all the time. Imbalanced load is a very common use case in practice.
```
Comment (Hoverbear) [inline text/2019-03-04-hibernate-raft.md]:
```suggestion
working all the time. Unbalanced load is a very common use case in practice.
```

File: text/2019-03-04-hibernate-raft.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+If during that time, a new node is added to the cluster, because of 1, leader
+will keep ticking until the conf change is complete and new node has caught up
+all the logs. If a node is removed, and the node is isolated, we need another
+mechanism to make it removed. This will be discussed when talking about
+`CheckPeerStaleState`.
+
+### RaftLogGc
+
+Apparently, only leader needs to tick. And when raft tick is stopped, this
```
Comment (Hoverbear) [inline text/2019-03-04-hibernate-raft.md]:
```suggestion
Only leader needs to tick. And when raft tick is stopped, this
```

Let's make sure the RFC isn't working on assumptions. Is it true that only the leader needs to tick or not?

File: text/2019-03-04-hibernate-raft.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+### CheckMerge
+
+It has been implemented in the lazy way already that it's only scheduled when
+a `PreMerge` is proposed while `CommitMerge` is not applied.
+
+### CheckPeerStaleState
+
+Hibernation means it should not be interrupted only when it's really
+necessary. This tick is deprecated completely. We have two ways to clean up
```
Comment (Hoverbear) [inline text/2019-03-04-hibernate-raft.md]: Is it worthwhile to investigate removing it completely?

**Comment by nolouch** [comment]
@BusyJay OK, I will try to.

**Comment by BusyJay** [comment]
PTAL

File: text/2019-03-04-hibernate-raft.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+They start campaigns in the end. And their campaign messages will wake the
+whole cluster up because of 5. Note that in such case, new leader will not
+stop ticking even it is elected and applies the new entry because there is
+still one out of service node, which is the old leader. Rule 4 and 8 can be
+more aggressive like if it detects that it has not received messages from
+leader for more than election_timeout, it can start campaign immediately.
+
+If during that time, follower is out of service and recovered before any
+proposals, it will start ticking because of 3. To prevent it from disturbing
+the leader, PreVote needs to be enabled. And leader will eventually be woken
```
Comment (Hoverbear) [inline text/2019-03-04-hibernate-raft.md]:
```suggestion
the leader, PreVote needs to be enabled. This way the leader will eventually be woken
```

File: text/2019-03-04-hibernate-raft.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+up and broadcasting heartbeat until all followers have been calmed down and go
+to sleep as 7 indicates. If follower is recovered after proposals, leader has
+been woken up already, and the whole cluster will be made sure to be up to
+date.
+
+If during that time, a new node is added to the cluster, because of 1, leader
```
Comment (Hoverbear) [inline text/2019-03-04-hibernate-raft.md]:
```suggestion
During this time, if a new node is added to the cluster, because of 1, leader
```

**Comment by hicqu** [comment]
LGTM.

File: text/2019-03-04-hibernate-raft.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+### PdHeartbeat
+
+This is also similar to `RaftLogGc`. However, `PdHeartbeat` must tick when
+there are any read proposals to let pd rebalance correctly.
+
+### CheckMerge
+
+It has been implemented in the lazy way already that it's only scheduled when
+a `PreMerge` is proposed while `CommitMerge` is not applied.
```
Comment (overvenus) [inline text/2019-03-04-hibernate-raft.md]: What is `PreMerge`?

File: text/2019-03-04-hibernate-raft.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+checking heartbeats. To make sure followers check the leader heartbeat
+reliablely, we can add a intermediate state to group state as PreChaos. If a
+follower is Idle, then change it to PreChaos; if it's PreChaos, then change it
+to Chaos. So it takes two ticks for follower to actually start campaign while
+it only takes one tick for leader to broadcast heartbeats. So followers should
+generally stay calm if network and clock is good.
+
+## Drawbacks
+
+When a leader is down, it may take a longer time to recover than before.
```
Comment (overvenus) [inline text/2019-03-04-hibernate-raft.md]: What about transit to `Chaos` when a follower receives a client request?

Comment (BusyJay) [reply text/2019-03-04-hibernate-raft.md]: It's part of the design, see L44. But client still needs to wait at least a round of election.

**Comment by BusyJay** [comment]
PTAL

**Pull request merged**